config/azure.go

```diff
@@ -2,13 +2,109 @@ package config
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"net/url"
+	"strings"
 
+	"github.com/databricks/databricks-sdk-go/apierr"
 	"github.com/databricks/databricks-sdk-go/httpclient"
 	"github.com/databricks/databricks-sdk-go/logger"
 	"golang.org/x/oauth2"
 )
 
+type tokenError struct {
+	message string
+	err     *httpclient.HttpError
+}
+
+func (e *tokenError) Error() string {
+	return e.message
+}
+
+func (e *tokenError) Unwrap() error {
+	sdkErr, ok := apierr.ByStatusCode(e.err.StatusCode)
+	if ok {
+		// this is how we distinguish between bad requests and permission denies
+		return sdkErr
+	}
+	return e.err
+}
+
+func (c *Config) mapAzureError(defaultErr *httpclient.HttpError) error {
+	env := c.Environment()
+	switch defaultErr.Request.Host {
+	case c.hostOrEmpty(env.AzureActiveDirectoryEndpoint()):
+		return c.mapAzureActiveDirectoryError(defaultErr)
+	case c.hostOrEmpty(env.AzureResourceManagerEndpoint()):
+		return c.mapAzureResourceManagerError(defaultErr)
+	default:
+		// Azure MSI endpoint returns not so typed error bodies: `404 page not found`
+		return &tokenError{
+			message: defaultErr.Message,
+			err:     defaultErr,
+		}
+	}
+}
+
+func (c *Config) hostOrEmpty(endpoint string) string {
+	parsedURL, err := url.Parse(endpoint)
+	if err != nil {
+		return ""
+	}
+	return parsedURL.Host
+}
+
+type azureActiveDirectoryErrorResponse struct {
+	CorrelationID    string `json:"correlation_id,omitempty"`
+	ErrorType        string `json:"error"`
+	ErrorCodes       []int  `json:"error_codes"`
+	ErrorDescription string `json:"error_description"`
+	ErrorURI         string `json:"error_uri"`
+}
+
+func (c *Config) mapAzureActiveDirectoryError(defaultErr *httpclient.HttpError) error {
+	var aadError azureActiveDirectoryErrorResponse
+	err := json.Unmarshal([]byte(defaultErr.Message), &aadError)
+	if err != nil {
+		return defaultErr
+	}
+	// remove rather explicit error description, as we're adding a link
+	// in the error rendering interface
+	msg, _, ok := strings.Cut(aadError.ErrorDescription, ". Trace ID")
```

Comment:
What information is being removed here? If there is debugging information that could be relevant to users, we may want to include it anyways?

Reply:
Trace ID / Correlation ID. If user is doing debug logs - it'll appear there anyway (in pretty-printed JSON). The most useful portion of info to resolve the issue is in the ErrorLink, which we appent later.